sdk/python/lib/pulumi/provider/experimental/provider.py

```diff
@@ -0,0 +1,57 @@
+# Copyright 2025, Pulumi Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from pathlib import Path
+from typing import Optional
+
+from ...output import Inputs
+from ...resource import ResourceOptions
+from ..provider import ConstructResult, Provider
+from .analyzer import Analyzer
+from .metadata import Metadata
+from .schema import generate_schema
+
+
+class ComponentProvider(Provider):
+    """
+    ComponentProvider is a Pulumi provider that finds components from Python
+    source code and infers a schema.
+    """
+
+    path: Path
+    """The path to the Python source code."""
+    metadata: Metadata
+    """The metadata for the provider, such as the name and version."""
+
+    def __init__(self, metadata: Metadata, path: Path) -> None:
+        self.path = path
+        self.metadata = metadata
+        self.analyzer = Analyzer(self.metadata)
+        self.components = self.analyzer.analyze(self.path)
+        schema = generate_schema(metadata, self.components)
+        super().__init__(metadata.version, json.dumps(schema.to_json()))
+
+    def construct(
+        self,
+        name: str,
+        resource_type: str,
+        inputs: Inputs,
+        options: Optional[ResourceOptions] = None,
+    ) -> ConstructResult:
+        component_name = resource_type.split(":")[-1]
```

Comment:
Should we validate all parts of `resource_type`?

Reply:
Good call, which makes me wonder how we want to handle modules different than `index`.

Reply:
Added here https://github.com/pulumi/pulumi/pull/18401/commits/1b96145f2d93a913357d30c3e52ee00413a4d5a7